src/metabase/domain_entities/malli.cljc

```diff
@@ -19,47 +19,113 @@
-  [schema]
-  (mut/update-properties schema assoc :bean/opaque true))
+     (mu/defn ^:export with-cache-ttl :- Question
+       \"Updater for [:card :cache-ttl].\"
+       [obj :- Question new-value :- number?]
+       (assoc-in obj [:card :cache-ttl] new-value))
+     ```
+
+     You provide the schema for the parent object; the macro will examine that schema to
+     determine the schema for the field being fetched or updated.
+
+     The updater's name gets prefixed with `with-`."
+     [schema & specs]
+     (let [parts        (partition 2 specs)]
+       `(do
+          ~@(mapcat (fn [[sym path]]
+                      (let [in-sym  (vary-meta (symbol (str "->" (name sym)))
+                                               assoc :private true)
+                            out-sym (vary-meta (symbol (str (name sym) "->"))
+                                               assoc :private true)]
+                        [;; Getter
+                         `(clojure.core/defn
+                            ~(vary-meta sym assoc :export true)
+                            ~(str "Accessor for `" path "`.")
+                            [obj#]
+                            (get-in obj# ~path))
+
+                         ;; Incoming converter for the replacement value.
+                         `(def ~in-sym
+                            ~(macros/case
+                               :cljs (macros/case
+                                       :cljs `(-> ~schema
+                                                  (metabase.domain-entities.malli/schema-for-path ~path)
+                                                  metabase.domain-entities.converters/incoming)
+                                       :clj  `identity)))
+
+                         ;; Setter
+                         `(clojure.core/defn
+                            ~(vary-meta (symbol (str "with-" (name sym)))
+                                        assoc :export true)
+                            ~(str "Updater for `" path "`.")
+                            [obj# new-value#]
+                            (assoc-in obj# ~path (~in-sym new-value#)))
+
+                         ;; JS converter
+                         (macros/case :cljs
+                           `(def ~out-sym
+                              (metabase.domain-entities.converters/outgoing
+                                (metabase.domain-entities.malli/schema-for-path ~schema ~path))))
+
+                         ;; JS-returning getter
+                         (macros/case :cljs
+                           `(clojure.core/defn
+                              ~(vary-meta (symbol (str (name sym) "-js"))
+                                          assoc :export true)
+                              ~(str "Fetches `" path "` and converts it to plain JS.")
+                              [obj#]
+                              (~out-sym (~sym obj#))))]))
+                    parts)))))
```

Comment:
A mega macro like this is way too hard to read. Break it out into smaller macros. This is how I would do it:

```clj
(defmacro -define-getter [sym path]
  `(clojure.core/defn
     ~(vary-meta sym assoc :export true)
     ~(str "Accessor for `" path "`.")
     [obj#]
     (get-in obj# ~path)))

(defmacro -define-converter
  "Incoming converter for the replacement value."
  [schema path in-sym]
  `(def ~in-sym
     ~(macros/case
        :cljs (macros/case
                :cljs `(-> ~schema
                           (metabase.domain-entities.malli/schema-for-path ~path)
                           metabase.domain-entities.converters/incoming)
                :clj  `identity))))

(defmacro -define-setter
  [sym path in-sym]
  `(clojure.core/defn
     ~(vary-meta (symbol (str "with-" (name sym)))
                 assoc :export true)
     ~(str "Updater for `" path "`.")
     [obj# new-value#]
     (assoc-in obj# ~path (~in-sym new-value#))))

(defmacro -define-js-converter
  [schema path out-sym]
  (macros/case
    :cljs
    `(def ~out-sym
       (metabase.domain-entities.converters/outgoing
        (metabase.domain-entities.malli/schema-for-path ~schema ~path)))))

(defmacro -define-js-returning-getter
  [sym path out-sym]
  (macros/case
    :cljs
    `(clojure.core/defn
       ~(vary-meta (symbol (str (name sym) "-js"))
                   assoc :export true)
       ~(str "Fetches `" path "` and converts it to plain JS.")
       [obj#]
       (~out-sym (~sym obj#)))))

(defmacro -define-getter-and-setter
  [schema sym path]
  (let [in-sym  (vary-meta (symbol (str "->" (name sym)))
                           assoc :private true)
        out-sym (vary-meta (symbol (str (name sym) "->"))
                           assoc :private true)]
    `(do
       (-define-getter ~sym ~path)
       (-define-converter ~schema ~path ~in-sym)
       (-define-setter ~sym ~path ~in-sym)
       (-define-js-converter ~schema ~path ~out-sym)
       (-define-js-returning-getter ~sym ~path ~out-sym))))

(defmacro define-getters-and-setters
  "Generates an accessor (`get-in`) and updater (`assoc-in`) for each specified path.
     For example:
     ```
     (def-getters-and-setters Question
       dataset-query [:card :dataset-query]
       cache-ttl     [:card :cache-ttl])
     ```
     will generate:
     ```
     (mu/defn ^:export dataset-query :- DatasetQuery
       \"Accessor for [:card :dataset-query].\"
       [obj :- Question]
       (get-in obj [:card :dataset-query]))
     (mu/defn ^:export with-dataset-query :- Question
       \"Updater for [:card :dataset-query].\"
       [obj :- Question new-value :- DatasetQuery]
       (assoc-in obj [:card :dataset-query] new-value))
     (mu/defn ^:export cache-ttl :- [:maybe number?]
       \"Accessor for [:card :cache-ttl].\"
       [obj :- Question]
       (get-in obj [:card :cache-ttl]))
     (mu/defn ^:export with-cache-ttl :- Question
       \"Updater for [:card :cache-ttl].\"
       [obj :- Question new-value :- number?]
       (assoc-in obj [:card :cache-ttl] new-value))
     ```
     You provide the schema for the parent object; the macro will examine that schema to
     determine the schema for the field being fetched or updated.
     The updater's name gets prefixed with `with-`."
  [schema sym path & more]
  `(do
     (-define-getter-and-setter ~schema ~sym ~path)
     ~(when (seq more)
        `(define-getters-and-setters ~schema ~@more))))
```

Notes:

1. Traditionally you only use `def` in macros when it's one word, e.g. `defparrot`, but if it's multiple words you use `define`, e.g. `define-noisy-parrot`
2. Since you can't make stuff in the macroexpansion private I prefixed the impl macros with `-` to make it more obvious to people that they're not supposed to be using them directly
3. `define-getters-and-setters` can expand to one `-define-getter-and-setter` macro for each of the pair of `sym` and `path`. Then you can have a much simpler `-define-getter-and-setter` macro that only has to deal with a fixed numner of args. While you **could** use `partition-all` to split the args it's actually even easier to do recursive macro expansions using `define-getter-and-setter` with the extra pairs until you run out of them
4. Instead of `-define-getter-and-setter` defining 5 things itself, it can expand to a series of 5 small macros that each expand to the specific thing you're defining.

A series of smaller macros like this is a lot easier to read I think, and if you're using CIDER or another tool you can easily see the macroexpansion one step at a time and not be blasted in the face with it all being expanded at once